pkg/render/metallb_config.go

```diff
@@ -0,0 +1,48 @@
+package render
+
+// This file contains the structures stolen from metallb config
+// package to ease the rendering
+
+type configFile struct {
+	Peers          []peer            `yaml:"peers,omitempty"`
+	BGPCommunities map[string]string `yaml:"bgp-communities,omitempty"` // TODO this is missing from crds
+	Pools          []addressPool     `yaml:"address-pools,omitempty"`
+}
+
+type peer struct {
+	MyASN         uint32         `yaml:"my-asn"`
+	ASN           uint32         `yaml:"peer-asn"`
+	Addr          string         `yaml:"peer-address"`
+	SrcAddr       string         `yaml:"source-address,omitempty"`
+	Port          uint16         `yaml:"peer-port,omitempty"`
+	HoldTime      string         `yaml:"hold-time,omitempty"`
+	RouterID      string         `yaml:"router-id,omitempty"`
+	NodeSelectors []nodeSelector `yaml:"node-selectors,omitempty"`
+	Password      string         `yaml:"password,omitempty"`
+}
+
+type nodeSelector struct {
+	MatchLabels      map[string]string      `yaml:"match-labels,omitempty"`
+	MatchExpressions []selectorRequirements `yaml:"match-expressions,omitempty"`
+}
+
+type selectorRequirements struct {
+	Key      string   `yaml:"key"`
+	Operator string   `yaml:"operator"`
+	Values   []string `yaml:"values"`
+}
+
+type addressPool struct {
+	Protocol          Proto
+	Name              string
+	Addresses         []string
+	AvoidBuggyIPs     bool               `yaml:"avoid-buggy-ips,omitempty"`
+	AutoAssign        *bool              `yaml:"auto-assign,omitempty"`
+	BGPAdvertisements []bgpAdvertisement `yaml:"bgp-advertisements,omitempty"`
+}
+
+type bgpAdvertisement struct {
+	AggregationLength *int `yaml:"aggregation-length"`
+	LocalPref         *uint32
+	Communities       []string
+}
```

Comment:
why redefine the types here all types.go is already have yaml object ?

Reply:
To match 1:1 the metallb configuration (these are all copied from metallb), so we marshal the very same object that metallb unmarshals

Reply:
Adding to @msherif1234 's question, isn't what defined in types.go already copied from metallb "by nature"? 
Couldn't we use metallbv1alpha1.BGPPeerSpec/AddressPoolSpec in the configFile struct and get the same results (and if not could it indicate that we didn't define the api correctly?) ?

Reply:
Ok, let me try to expand my reasoning.

The operator provides an abstraction over what metallb provides. It's a different API (otherwise, we would be exposing the same configmap). We have CRDs on one side, and the input for MetalLB on the other.

Even if it's a 1:1 mapping between the items, here we decouple the external API from what we feed metallb with because they are two different domains. Also, note that the 1:1 mapping is only between the items, not between the api as a whole, because the operator handles different scattered CRs while metallb accepts one big configuration.

Note that the external API was already decoupled by the metallb configuration, it was just being done in the template code. We leverage the fact that we have the metallb structure available, so doing the translation is straightforward.

Reply:
@msherif1234 does this sound convincing?
@oribon approved the PR but I'd like to reach consensus.